### PR TITLE
Document line symbol layers properties

### LINKED ISSUE #3638: [FEATURE] New line symbol type: Hash line
Original commit: https://github.com/qgis/QGIS/commit/48d2a370570e43cd6c9dedf802ca4549d74f7af8 by nyalldawson

This line symbol type is designed to replicate the ArcGIS Hash Line
symbol layer type. It allows for a repeating line segment to be
drawn over the length of a feature, with a line-sub symbol used
to render each individual segment.

To reduce code duplication, this is heavily based off the current
line marker symbol layer, since the functionality is almost
identical (draw some sub symbol at some interval along a line).
Accordingly, I've split off QgsMarkerLineSymbolLayer to move
as much of the common functionality as possible to a new abstract
base class, so that only the actual marker/line segment rendering
occurs in the marker line/hash line subclasses.

This also gives the hash line all the existing placement options
permissible for marker lines -- e.g. first/last vertex, mid points,
regular intervals, etc.

The hash line length and angle can have data defined overrides,
which are evaluated per-line segment, allowing for the hash line
to change size and angle over the length of a single rendered
feature.

### LINKED ISSUE #3646: [FEATURE] Average line angles for marker and hashed line symbology
Original commit: https://github.com/qgis/QGIS/commit/426382c919a0082e86d22b63200bb9c39b065573 by nyalldawson

Previously, when marker or hash lines were rendered using interval
or center point place placement, the symbol angles were determined
by taking the exact line orientation at the position of the symbol.

This often leads to undesirable rendering effects, where little
jaggies or corners in lines which occur at the position of the
symbol cause the marker or hash line to be oriented at a very
different angle to what the eye expects to see.

With this new option, the angle is instead calculated by averaging
the line over a specified distance either side of the symbol. E.g.
averaging the line angle over 4mm means we take the points along
the line 2mm from either side of the symbol placement, and use these
instead to calculate the line angle for that symbol. This has the
effect of smoothing (or removing) any tiny local deviations from
the overall line direction, resulting in much nicer visual
orientation of marker or hash lines.

Like all symbol settings, the average angle smoothing distance
can be set using mm/pixels/map units/etc, and supports data-defined
values.

Closed rings also correctly consider wrapping around these average
angles from the start/end vertex.

(Sponsored by an anonymous corporate backer)

### LINKED ISSUE #3709: [needs-docs] Tweak custom dash button appearance
Original commit: https://github.com/qgis/QGIS/commit/8d180c1e8a51f66f8ba7558b98023f1c758d241d by nyalldawson

- remove "Change" label and replace with larger dash preview icon.
The "change" text is unnecessary and adds to dialog clutter, better
to use the space for a wider preview icon (especially given that
the previous narrow icon never really showed enough of the pattern
to be useful!)

- don't offset the line in the preview if the symbol has an offset
set

- respond correctly to dash pattern, line width unit changes, cap
style changes

- show a nice big preview tooltip on hover

### LINKED ISSUE #4111: [FEATURE] "Center of segment" placement mode for marker and hash line symbol layers
Original commit: https://github.com/qgis/QGIS/commit/2ca302999dfd7e1e68bcf881a731c2c66483ad9d by nyalldawson

Adds a new placement option to place markers/hash lines at the center
point of individual line segments

Refs #29785

### PR DESCRIPTION
**Pull request by DelazJ**
with screenshots
Fix #3638, fix #3646, fix #3709, fix #4111, ref #1738

### CODE REVIEW

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 1]
```diff
@@ -264,14 +271,52 @@ layer types:
...
-  central point or on every curve point. You can set an offset along the line
-  for the marker symbol, or offset the line itself. The :guilabel:`Rotate
-  marker` option allows you to set whether the marker symbol should follow the
-  line orientation or not.
+  layer symbol helps you representing flow maps.
+* **Geometry generator** (see :ref:`geometry_generator_symbol`)
+
+.. _marker_line_symbol:
+
+* **Marker line**: allows for a repeating :ref:`marker symbol
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
* **Marker line**: repeats a :ref:`marker symbol
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 2]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+  <vector_marker_symbols>` to be drawn over the length of a line symbol.
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
  <vector_marker_symbols>` over the length of a line.
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 3]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+
+  * The markers placement can be at a regular distance or based on the line
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
  * The markers can be placed with regular distances, or be based on the line
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 4]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+    geometry: first, last or each vertex, on the central point of the line
+    or of each segment, or on every curve point.
+  * The markers placement can also be offset along the line
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
  * The markers placement can also be given an offset along the line
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 5]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+  * The |checkbox| :guilabel:`Rotate marker` option allows you to set whether
+    each marker symbol should follow the line orientation or its own rotation.
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
    each marker symbol should be oriented relative to the line direction or not.
```

Comment (havatv) [reply source/docs/user_manual/style_library/symbol_selector.rst]: This is important.  The specified "rotation" for the marker symbol will be applied also if "rotate marker" is checked.

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 6]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+
+    Because a line is often a succession of segments of different directions,
+    the marker's rotation is calculated by averaging the line over a specified
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
    the rotation of the marker is calculated by averaging over a specified
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 7]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+    distance either side of the symbol. For example, setting the
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
    distance along the line on. For example, setting the
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 8]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+    :guilabel:`Average angle over` property to ``4mm`` means to take the points
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
    :guilabel:`Average angle over` property to ``4mm`` means that the two points
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 9]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+    along the line ``2mm`` from either side of the symbol placement, and use
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
    along the line that are ``2mm`` before and after the symbol placement are used
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 10]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+    these instead to calculate the line angle for that symbol.
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
    to calculate the line angle for that marker symbol.
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 11]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+    This has the effect of smoothing (or removing) any tiny local deviations
+    from the overall line direction, resulting in much nicer visual orientation
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
    from the overall line direction, resulting in much nicer visual orientations
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 12]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+    of the marker line.
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
    of the marker line symbols.
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 13]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+  * The marker line can also be offset from the line itself.
+
+.. _hashed_line_symbol:
+
+* **Hashed line**: allows for a repeating line segment (a hash) to be drawn
+  over the length of a line symbol, with a line sub-symbol used to render each
+  individual segment. In other words, a hashed line is like a marker line in
+  which marker symbols are replaced with segments. As such, the hashed lines
+  have :ref:`same properties <marker_line_symbol>` as marker line symbols,
```
Comment (havatv) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
  have the :ref:`same properties <marker_line_symbol>` as marker line symbols,
```

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 14]
```diff
@@ -248,14 +250,19 @@ layer types:
...
-  * :guilabel:`Stroke style`
-  * :guilabel:`Join style`
+  .. _figure_simple_line_symbol:
+
+  .. figure:: img/simpleLineSymbol.png
+     :align: center
+
+     Designing a Simple Line Symbol
+
+  Along with common properties of the :ref:`simple marker symbol
```
[the remark this answers is not in the document]
Comment (DelazJ) [reply source/docs/user_manual/style_library/symbol_selector.rst]: Actually, a line does not show ALL the properties of a marker, just some of them. And here I wanted a way to introduce only the new ones.

File: source/docs/user_manual/style_library/symbol_selector.rst (hunk 3 above)
Comment (DelazJ) [reply source/docs/user_manual/style_library/symbol_selector.rst]: It's the marker placement that is based on the line geometry, not the marker itself

File: source/docs/user_manual/style_library/symbol_selector.rst (hunk 5 above)
Comment (DelazJ) [reply source/docs/user_manual/style_library/symbol_selector.rst]: indeed

File: source/docs/user_manual/style_library/symbol_selector.rst [hunk 15]
```diff
@@ -264,14 +271,52 @@ layer types:
...
+    along the line ``2mm`` from either side of the symbol placement, and use
+    these instead to calculate the line angle for that symbol.
+    This has the effect of smoothing (or removing) any tiny local deviations
+    from the overall line direction, resulting in much nicer visual orientation
+    of the marker line.
+  * The marker line can also be offset from the line itself.
+
+.. _hashed_line_symbol:
+
+* **Hashed line**: allows for a repeating line segment (a hash) to be drawn
```
Comment (DelazJ) [inline source/docs/user_manual/style_library/symbol_selector.rst]:
```suggestion
* **Hashed line**: repeats a line segment (a hash)
```

File: source/docs/user_manual/style_library/symbol_selector.rst (hunk 3 above)
Comment (DelazJ) [reply source/docs/user_manual/style_library/symbol_selector.rst]: Not convinced this is better...

**Comment by DelazJ** [comment]
Thanks for the review @havatv

**Comment by DelazJ** [comment]
Unrelated failure

**Pull request merged**